fixes cache time handling in apt module

cloudnull opened the pull request:
This change is in response to issue #1497 where the apt module would not
properly updating the apt cache in some situations and never returned a state
change on cache update when the module was used without or without an item to
be installed or upgraded.

The change simply allows the apt module to update the cache when
`update_cache` option is used without or without a set `cache_valid_time`. If
`cache_valid_time` is set and the on disk mtime for apt cache is ">" the
provided amount of seconds, which now has a default of 0, the apt cache will
be updated. Additionally if no upgrade, package, or deb is installed or
changed but the apt cache is updated the module will return a changed state
which will help users to know that the state of the environment has changed
due to a task operation, even if it was only an apt cahce update.

Signed-off By: Kevin Carter kevin.carter@rackspace.com

fixes #1497

odyssey4me commented on packaging/os/apt.py:
```diff
@@ -536,12 +536,40 @@ def upgrade(m, mode="yes", force=False, default_release=None,
     m.exit_json(changed=True, msg=out, stdout=out, stderr=err)
 
+def get_cache_mtime():
+    """Retrun mtime of a valid apt cache file.
```
'return', perhaps instead of 'retrun'?

cloudnull replied:
Done.

gregdek commented:
@mgwilliams Please review this PR to make sure it adheres to the following guidelines:

http://docs.ansible.com/developing_modules.html#module-checklist

If it passes these guidelines, and if you believe it’s a good PR otherwise, please give it a +1; at that point, we will add the “priority_reviewed” flag to this PR, which will move it to the top of our queue for potential inclusion.

bcoca commented on packaging/os/apt.py:
```diff
@@ -596,39 +624,27 @@ def main():
+        # Cache valid time is default 0, which will update the cache if
+        #  needed and `update_cache` was set to true
+        updated_cache = False
+        if p['update_cache'] or p['cache_valid_time']:
```
this changes the behavior and now setting a cache_valid_time implies updated_cache=True

cloudnull replied:
@bcoca This does change the behavior by allowing the user to simply set a `cache_valid_time` which would cause the update to happen if the time was older than what was specificed. If you'd think it best I can remove the `or` which would set it back to the old behavior.

cloudnull commented:
@gregdek & @bcoca I just noticed the _needs_revision_ tag so I've updated the method to remove the `or` in the cache update conditional which should keep the old behavior. Let me know if you find anything else.

gregdek commented:
Putting back into core_review for @bcoca

cloudnull commented:
:bump: I'm curious if there's anything here that I need to work on or should update for this patch?

bcoca commented:
@cloudnull its on my list to review and test, i just have a lot on that list.

cloudnull commented:
@bcoca no worries. I just got back to the office and was checking in on all the things i have in flight. Thanks for the update! I hope all is well.

cloudnull commented:
Rebased off of the head of devel because there were merge conflicts.

gregdek commented:
Thanks @cloudnull for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

cloudnull commented:
I've gone and rebased this change set. Thank you for the initial review. If there's anything else I need to update or you'd like to see please let me know.

cloudnull commented:
ready_for_review

gregdek commented:
Thanks @cloudnull for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

gregdek commented:
Thanks @cloudnull for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

cloudnull commented:
TYVM @gregdek please let me know if I can help out in any way.

gregdek commented:
Thanks @cloudnull for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

cloudnull commented:
ready_for_review -- Thank you for the update @gregdek I've rebased the changeset off of the head of the current devel branch. Please let me know if anything else needs to be updated.

gregdek commented:
Thanks @cloudnull for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

cloudnull commented:
The last push was to update the module to resolve recent changes causing a merge conflict in this PR.

cloudnull commented:
ready_for_review

gregdek commented:
Thanks @cloudnull for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

cloudnull commented:
@gregdek bump, and idea when this might be able to be reviewed?

gregdek commented:
Still up to the core team. Ping @ansible/core.

gregdek commented:
Also worth perhaps bringing it up at the Core Team Meeting, which is every Tuesday 1900 UTC and Thursday 1500 UTC on #ansible-meeting on freenode.irc.net. 

https://github.com/ansible/community/blob/master/MEETINGS.md

cloudnull commented:
@gregdek I've added the meeting to my calendar. Thanks again for the update.

senorsmile commented:
The bug that this PR fixes just hit me.  What's the status on merging?

cloudnull commented:
@senorsmile ++ I'd love to see this get in for "2.2" though we're waiting on core review.

resmo commented on packaging/os/apt.py:
```diff
@@ -682,47 +705,34 @@ def main():
                         pass
-                else:
-                    #out of retries, pass on the exception
-                    raise
```
Why not raise the exception the cache update went wrong?

cloudnull replied:
Good point, I'll add that back.

resmo commented:
I made a code review and the code looks straight forward and looks more readable to me. I am heading to testing it.

cloudnull commented:
@resmo thanks for the review, i've updated the PR based on your feedback. let me know how it goes.

cloudnull committed:
This change is in response to issue #1497 where the apt module would not

properly updating the apt cache in some situations and never returned a state
change on cache update when the module was used without or without an item to
be installed or upgraded.

The change simply allows the apt module to update the cache when
update_cache option is used without or without a set cache_valid_time. If
cache_valid_time is set and the on disk mtime for apt cache is ">" the
provided amount of seconds, which now has a default of 0, the apt cache will
be updated. Additionally if no upgrade, package, or deb is installed or
changed but the apt cache is updated the module will return a changed state
which will help users to know that the state of the environment has changed
due to a task operation, even if it was only an apt cahce update.

Signed-off By: Kevin Carter kevin.carter@rackspace.com
fixes #1497

Signed-off-by: Kevin Carter <kevin.carter@rackspace.com>

cloudnull commented:
The last PR was a rebase off the HEAD of `devel` because CI checks were failing when rebasing this PR.

cloudnull commented:
fixed syntax problem in py3.4/5 tests:

``` python-traceback
The command "python3.4 -m compileall -fq ." exited with 1.
0.83s$ python3.5 -m compileall -fq .
*** Error compiling './packaging/os/apt.py'...
  File "./packaging/os/apt.py", line 465
    except Exception, e:
                    ^
SyntaxError: invalid syntax
```

sivel commented:
To properly handle exceptions for both py2.4 through py3, see http://docs.ansible.com/ansible/developing_modules_python3.html#exceptions

cloudnull commented:
Ah.. I forgot about 2.4 support... :| PR inbound.

cloudnull commented:
@resmo @sivel -- IDK whats up with **shippable** all but 1 test passed and **travis-ci** seems to be happy. Any pointers?

mattclay commented:
@cloudnull There was a transient failure on Shippable due to an issue connecting to Docker Hub. I've restarted the build.

gregdek commented:
Thanks @cloudnull for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

resmo commented:
@cloudnull Hi Kevin, do you have time to rebase and resolve the conflicts? It is on my list for looking at it before release of 2.2.

cloudnull commented:
@resmo I've rebased this PR. let me know if you need anything else. Thanks again for taking the time to review this.

gregdek commented:
@cloudnull A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

cloudnull commented:
ready_for_review

gregdek commented:
Thanks @cloudnull for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

resmo commented on packaging/os/apt.py:
```diff
@@ -772,46 +797,36 @@ def main():
                         pass
                 else:
-                    module.fail_json(msg='Failed to update apt cache.')
+                    raise SystemError('apt failed to update')
```
I would prefer the `module.fail_json` as it was.

cloudnull replied:
Done.

resmo reviewed (changes requested):
Tested on Debian 7 and 8, worked as expected. I would only like to see the use of fail_json if cache update failed

Commit:
Update apt.py

resmo commented:
LGTM

gregdek commented:
Thanks again to @cloudnull for this PR. This PR was reviewed by an Ansible member. Marking for inclusion.

However, we would still appreciate if the maintainer(s) @ansible could look over it.

[This message brought to you by your friendly Ansibull-bot.]

resmo commented:
Thanks!

sYnfo commented:
@greenbrian I think https://github.com/ansible/ansible-modules-core/blob/845cc1a531610f2db9639393acc7c1663235716e/packaging/os/apt.py#L911 might contain a bug. `retvals` does not necessarily contain `'changed'`, in particular if `install` (https://github.com/ansible/ansible-modules-core/blob/845cc1a531610f2db9639393acc7c1663235716e/packaging/os/apt.py#L439) fails it will certainly not.
